## Tasca S3 T02

In [1]:
import numpy as np
import pandas as pd

**Redueix la dimensió del dataset de manera aleatòria per tal d'obtenir un dataset de només 200.000 registres.**

### Exercici 1

Descarrega el data set Airlines Delay: Airline on-time statistics and delay causes i carrega’l a un Pandas Dataframe. Explora les dades que conté, explica breument quines variables hi ha i queda’t únicament amb les columnes que consideris rellevants. Justifica la teva elecció.

In [2]:
df = pd.read_csv("dataset/DelayedFlights.csv", index_col=0)

Obtenemos informacion basica del tipo de datos que contiene el dataset con la funcion *info()*.

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1936758 entries, 0 to 7009727
Data columns (total 29 columns):
 #   Column             Dtype  
---  ------             -----  
 0   Year               int64  
 1   Month              int64  
 2   DayofMonth         int64  
 3   DayOfWeek          int64  
 4   DepTime            float64
 5   CRSDepTime         int64  
 6   ArrTime            float64
 7   CRSArrTime         int64  
 8   UniqueCarrier      object 
 9   FlightNum          int64  
 10  TailNum            object 
 11  ActualElapsedTime  float64
 12  CRSElapsedTime     float64
 13  AirTime            float64
 14  ArrDelay           float64
 15  DepDelay           float64
 16  Origin             object 
 17  Dest               object 
 18  Distance           int64  
 19  TaxiIn             float64
 20  TaxiOut            float64
 21  Cancelled          int64  
 22  CancellationCode   object 
 23  Diverted           int64  
 24  CarrierDelay       float64
 25  WeatherDelay      

Esta base de datos está compuesta por las siguientes variables:
* **Year**  2008
* **Month** 1-12
* **DayofMonth** 1-31
* **DayOfWeek** 1 (Monday) - 7 (Sunday)
* **DepTime** actual departure time (local, hhmm)
* **CRSDepTime** scheduled departure time (local, hhmm)
* **ArrTime actual** arrival time (local, hhmm)
* **CRSArrTime** scheduled arrival time (local, hhmm)
* **UniqueCarrier** airline code
* **FlightNum** flight number
* **TailNum** plane tail number: aircraft registration, unique aircraft identifier
* **ActualElapsedTime** in minutes
* **CRSElapsedTime** in minutes
* **AirTime** in minutes
* **ArrDelay** arrival delay, in minutes: A flight is counted as "on time" if it operated less than 15 minutes later the scheduled time shown in the carriers' Computerized Reservations Systems (CRS).
* **DepDelay** departure delay, in minutes
* **Origin** origin IATA airport code
* **Dest** destination IATA airport code
* **Distance** in miles
* **TaxiIn** taxi in time, in minutes
* **TaxiOut** taxi out time in minutes
* **Cancelled** 1 = yes, 0 = no
* **CancellationCode** reason for cancellation (A = carrier, B = weather, C = NAS, D = security)
* **Diverted** 1 = yes, 0 = no
* **CarrierDelay** in minutes: Carrier delay is within the control of the air carrier. Examples of occurrences that may determine carrier delay are: aircraft cleaning, aircraft damage, awaiting the arrival of connecting passengers or crew, baggage, bird strike, cargo loading, catering, computer, outage-carrier equipment, crew legality (pilot or attendant rest), damage by hazardous goods, engineering inspection, fueling, handling disabled passengers, late crew, lavatory servicing, maintenance, oversales, potable water servicing, removal of unruly passenger, slow boarding or seating, stowing carry-on baggage, weight and balance delays.
* **WeatherDelay** in minutes: Weather delay is caused by extreme or hazardous weather conditions that are forecasted or manifest themselves on point of departure, enroute, or on point of arrival.
* **NASDelay** in minutes: Delay that is within the control of the National Airspace System (NAS) may include: non-extreme weather conditions, airport operations, heavy traffic volume, air traffic control, etc.
* **SecurityDelay** in minutes: Security delay is caused by evacuation of a terminal or concourse, re-boarding of aircraft because of security breach, inoperative screening equipment and/or long lines in excess of 29 minutes at screening areas.
* **LateAircraftDelay** in minutes: Arrival delay at an airport due to the late arrival of the same aircraft at a previous airport. The ripple effect of an earlier delay at downstream airports is referred to as delay propagation.

Para seleccionar los datos de nuestro interés, eliminamos las columnas no deseadas. Utilizaremos la función *drop()*

In [4]:
df = df.drop(['Year'], axis=1)
df = df.drop(['DepTime'], axis=1) 
df = df.drop(['DepDelay'], axis=1)
df = df.drop(['ArrTime'], axis=1)
df = df.drop(['CRSArrTime'], axis=1)
df = df.drop(['ActualElapsedTime'], axis=1)
df = df.drop(['CRSElapsedTime'], axis=1)
df = df.drop(['FlightNum'], axis=1)
df = df.drop(['TailNum'], axis=1)
df = df.drop(['TaxiIn'], axis=1)
df = df.drop(['TaxiOut'], axis=1)

Utilizamos la función *head()* para visualizar las primeras filas de datos de la estructura del dataset

In [5]:
df.head(5)

Month  DayofMonth  DayOfWeek  CRSDepTime UniqueCarrier  AirTime  ArrDelay  \
0      1           3          4        1955            WN    116.0     -14.0   
1      1           3          4         735            WN    113.0       2.0   
2      1           3          4         620            WN     76.0      14.0   
4      1           3          4        1755            WN     77.0      34.0   
5      1           3          4        1915            WN     87.0      11.0   

  Origin Dest  Distance  Cancelled CancellationCode  Diverted  CarrierDelay  \
0    IAD  TPA       810          0                N         0           NaN   
1    IAD  TPA       810          0                N         0           NaN   
2    IND  BWI       515          0                N         0           NaN   
4    IND  BWI       515          0                N         0           2.0   
5    IND  JAX       688          0                N         0           NaN   

   WeatherDelay  NASDelay  SecurityDelay  LateAircraftDelay  
0           NaN       NaN            NaN                NaN  
1           NaN       NaN            NaN                NaN  
2           NaN       NaN            NaN                NaN  
4           0.0       0.0            0.0               32.0  
5           NaN       NaN            NaN                NaN

In [8]:
data.shape

(1223399, 17)

### Exercici 2

* Resumeix estadísticament el dataset i les columnes d’interès. Fes una anàlisi estadístic del que consideris rellevant.

In [6]:
round(df.describe(),2)

Month  DayofMonth   DayOfWeek  CRSDepTime     AirTime    ArrDelay  \
count  1936758.00  1936758.00  1936758.00  1936758.00  1928371.00  1928371.00   
mean         6.11       15.75        3.98     1467.47      108.28       42.20   
std          3.48        8.78        2.00      424.77       68.64       56.78   
min          1.00        1.00        1.00        0.00        0.00     -109.00   
25%          3.00        8.00        2.00     1135.00       58.00        9.00   
50%          6.00       16.00        4.00     1510.00       90.00       24.00   
75%          9.00       23.00        6.00     1815.00      137.00       56.00   
max         12.00       31.00        7.00     2359.00     1091.00     2461.00   

         Distance   Cancelled    Diverted  CarrierDelay  WeatherDelay  \
count  1936758.00  1936758.00  1936758.00    1247488.00    1247488.00   
mean       765.69        0.00        0.00         19.18          3.70   
std        574.48        0.02        0.06         43.55         21.49   
min         11.00        0.00        0.00          0.00          0.00   
25%        338.00        0.00        0.00          0.00          0.00   
50%        606.00        0.00        0.00          2.00          0.00   
75%        998.00        0.00        0.00         21.00          0.00   
max       4962.00        1.00        1.00       2436.00       1352.00   

         NASDelay  SecurityDelay  LateAircraftDelay  
count  1247488.00     1247488.00         1247488.00  
mean        15.02           0.09              25.30  
std         33.83           2.02              42.05  
min          0.00           0.00               0.00  
25%          0.00           0.00               0.00  
50%          2.00           0.00               8.00  
75%         15.00           0.00              33.00  
max       1357.00         392.00            1316.00

* Troba quantes dades faltants hi ha per columna.

In [7]:
df.isna().sum()

Month                     0
DayofMonth                0
DayOfWeek                 0
CRSDepTime                0
UniqueCarrier             0
AirTime                8387
ArrDelay               8387
Origin                    0
Dest                      0
Distance                  0
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64

Se observar que 13 columnas tienen datos faltantes en el dataset original.  
Principalmente ello se deba a que si un vuelo es cancelado o desviado no acumule datos de retraso en alguna de las columnas de interés.

* Crea columnes noves:

    * Si ha arribat tard o no...

Se ha creado un condicional mediante el cual podemos saber el status del vuelo. Por lo general, un vuelo se cuenta como "a tiempo" si operó en menos de 15 minutos después de la hora programada que se muestra en los Sistemas de Reservas Computarizados (CRS) de las aerolíneas.

In [8]:
#seleccionamos los datos de interes 
df.loc[df['ArrDelay'] <= 15, 'Status'] = "On time"
df.loc[df['ArrDelay'] > 15, 'Status'] = "Delayed"
df.loc[df['Diverted'] == 1,  'Status'] = "Diverted"
df.loc[df['Cancelled'] == 1, 'Status'] = "Cancelled"

#eliminamos los datos no deseados
df = df.drop(['Diverted'], axis=1)
df = df.drop(['Cancelled'], axis=1)

#Filtramos para seleccionar los vuelos que acumulan retraso
data = df[df.Status != "On time"]

    * Velocitat mitjana del vol

La velocidad media de vuelo se definie como **V = Distance / AirTime**  
*Nota: se debe tener en cuenta las medidas de cada una de las variables*

In [9]:
data['Distance_km'] = data['Distance'] * 1.60934 #pasamos las unidades a Km
data['Vmean'] = data['Distance_km'] / (data['AirTime']/60) #creamos una nueva columna con la velocidad media [km/h]

C:\Users\maria\AppData\Local\Temp\ipykernel_20220\4092030889.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Distance_km'] = data['Distance'] * 1.60934 #pasamos las unidades a Km
C:\Users\maria\AppData\Local\Temp\ipykernel_20220\4092030889.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data['Vmean'] = data['Distance_km'] / (data['AirTime']/60) #creamos una nueva columna con la velocidad media [km/h]


In [10]:
data.head(5)

Month  DayofMonth  DayOfWeek  CRSDepTime UniqueCarrier  AirTime  ArrDelay  \
4       1           3          4        1755            WN     77.0      34.0   
6       1           3          4        1830            WN    230.0      57.0   
11      1           3          4        1510            WN    107.0      80.0   
18      1           3          4        1255            WN    110.0      16.0   
19      1           3          4        1325            WN     49.0      37.0   

   Origin Dest  Distance CancellationCode  CarrierDelay  WeatherDelay  \
4     IND  BWI       515                N           2.0           0.0   
6     IND  LAS      1591                N          10.0           0.0   
11    IND  MCO       828                N           8.0           0.0   
18    IND  TPA       838                N           0.0           0.0   
19    ISP  BWI       220                N          12.0           0.0   

    NASDelay  SecurityDelay  LateAircraftDelay   Status  Distance_km  \
4        0.0            0.0               32.0  Delayed    828.81010   
6        0.0            0.0               47.0  Delayed   2560.45994   
11       0.0            0.0               72.0  Delayed   1332.53352   
18       0.0            0.0               16.0  Delayed   1348.62692   
19       0.0            0.0               25.0  Delayed    354.05480   

         Vmean  
4   645.826052  
6   667.946071  
11  747.215058  
18  735.614684  
19  433.536490

* Fes una taula de les aerolínies amb més endarreriments acumulats.

In [11]:
#Top 5 de aerolinias con más retrasos acumulados
data.UniqueCarrier.value_counts().head(5)

WN    197825
AA    130356
MQ     95823
UA     93877
OO     87272
Name: UniqueCarrier, dtype: int64

La aerolinea que presenta mas retrasos es **Southwest Airlines(WN)**, seguido de **American Airlines (AA)** y **American Eagle Airlines (MQ)**.

* Quins són els vols més llargs? Busca les rutes més llargues.

In [12]:
max_ruta = data[data.Distance == data.Distance.max()]
max_ruta[['Origin','Dest','Distance']].drop_duplicates()

Origin Dest  Distance
566875    EWR  HNL      4962
584758    HNL  EWR      4962

La distancia total del vuelo más largo se encuentra entre Honolulu, Hawái (HNL) y Newark, Nueva Jersey (EWR) con un total de 4.962 millas. Un equivalente a 7.985 kilómetros.

In [13]:
data[['Origin','Dest','Distance']].sort_values('Distance',ascending=False).drop_duplicates().head(10)

Origin Dest  Distance
2971152    EWR  HNL      4962
6988170    HNL  EWR      4962
4828717    ATL  HNL      4502
6220927    HNL  ATL      4502
5037970    ORD  HNL      4243
2866988    HNL  ORD      4243
4454066    KOA  ORD      4213
3830608    ORD  OGG      4184
1018832    MSP  HNL      3972
438613     HNL  MSP      3972

* I els més endarrerits? Busca les que acumulen més retards.

In [14]:
delay = data[data.ArrDelay == data.ArrDelay.max()]
delay[['Origin','Dest','ArrDelay']]

Origin Dest  ArrDelay
1018798    HNL  MSP    2461.0

La ruta Honolulu, Hawái (HNL) y Mineápolis-Saint Paul, Minessota (MSP) és la que acumula el mayor retraso, 2461 min.

In [15]:
data[['Origin','Dest','ArrDelay']].sort_values('ArrDelay',ascending=False).drop_duplicates().head(10)

Origin Dest  ArrDelay
1018798    HNL  MSP    2461.0
2235378    CLT  MSP    2453.0
2832617    RSW  DTW    1951.0
3387883    LIT  DFW    1707.0
6857047    BOS  MSP    1655.0
5232546    OMA  MSP    1583.0
2232494    MOT  MSP    1542.0
527950     EGE  MIA    1525.0
4061361    SEA  MSP    1510.0
1634129    BNA  MEM    1490.0

* Aporta allò que consideris rellevant.

In [16]:
data.Status.value_counts()

Delayed      1215012
Diverted        7754
Cancelled        633
Name: Status, dtype: int64

De los vuelos que presentan retraso solo **633 (< 1%)** son cancelados y **7754 (< 1%)** son desviados

**CANCELLATION CODES:**
* A - Carrier Caused
* B - Weather 
* C - National Aviation System 
* D - Security

In [17]:
data.CancellationCode.value_counts()

N    1222766
B        307
A        246
C         80
Name: CancellationCode, dtype: int64

De los 633 vuelos cancelados: **307 fueron a causa de tiempo, 246 por carrier y 80 por seguridad.**

In [18]:
data.Origin.value_counts().head()

ORD    89370
ATL    85716
DFW    60667
DEN    45418
EWR    35681
Name: Origin, dtype: int64

Los tres aeropuertos que presentan más retrasos en vuelos son los de **Chicago (ORD)**, **Atlanta (ATL)** y **Dallas (DFW)** de 303 aeropuertos.

In [19]:
pd.crosstab(data.Month,data.Status)

Status  Cancelled  Delayed  Diverted
Month                               
1               0   117727       612
2               0   125591       909
3               0   127628       726
4               0    94917       481
5               0    92081       361
6               0   133275      1026
7               0   116394       774
8               0    97880       674
9               0    54062       188
10             59    55154       285
11             94    62012       321
12            480   138291      1397

**Diciembre y Junio** son respectivamente los meses que acumulan más retrasos. Por otra parte, se observa que los casos de cancelaciones se produjeron todos a finales de año.

In [20]:
data.DayOfWeek.value_counts().head(3)

5    210213
4    183032
1    183023
Name: DayOfWeek, dtype: int64

Siendo el **viernes (5)** el día con más retrasos de la semana

### Exercici 3

Exporta el dataset net i amb les noves columnes a Excel.

In [23]:
data.columns

Index(['Month', 'DayofMonth', 'DayOfWeek', 'CRSDepTime', 'UniqueCarrier',
       'AirTime', 'ArrDelay', 'Origin', 'Dest', 'Distance', 'CancellationCode',
       'CarrierDelay', 'WeatherDelay', 'NASDelay', 'SecurityDelay',
       'LateAircraftDelay', 'Status', 'Distance_km', 'Vmean'],
      dtype='object')

In [28]:
#seleccionamos con los datos relevantes a exportar
DelayedFlights = data[['Month', 'UniqueCarrier', 'ArrDelay', 'Origin', 'Dest', 'Distance','Status', 'Vmean']]

In [29]:
DelayedFlights.to_excel('dataset/DelayedFlights.xlsx')